Add test leg to the PR build to run libraries tests on iOS x64 Simulators

premun opened the pull request:
Forked from https://github.com/dotnet/runtime/pull/37057

Adds the iOS test leg that will execute selected BCL tests on an iOS x64 Simulator (13.5) using the `OSX.1015.Amd64.Open` queue which is our on-prem Mac Mini infrastructure (60 machines).

To prepare the Helix job, we are using the Arcade Helix SDK, more specifically the [XHarness Helix SDK](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/Readme.md) which uses the [XHarness tool](https://github.com/dotnet/xharness) to control the test run.

premun commented:
/azp run

premun committed:
Add XHarness Helix SDK support

premun committed:
Try to build mono libraries

premun committed:
Try to build mono libraries

premun commented on src/libraries/sendtohelix.proj:
```diff
@@ -97,7 +102,20 @@
-    <ItemGroup>
+    <ItemGroup Condition="'$(TargetOS)' == 'iOS' or '$(TargetOS)' == 'tvOS'">
+      <!-- Find all directories named *.app and strip the trailing slash -->
+      <_appBundleFiles Include="$(TestArchiveTestsRoot)**/*"/>
+      <_appBundleDirs Include="@(_appBundleFiles->'%(RootDir)%(Directory)'->Distinct())"/>
+      <XHarnessAppFolderToTest Include="@(_appBundleDirs->Replace(&quot;.app/&quot;, &quot;.app&quot;))"
+                               Condition="$([System.Text.RegularExpressions.Regex]::IsMatch(%(Identity), '\.app/?$'))">
```
**Help wanted:**
Can someone please let me know if there's a better way to find all directories ending with `.app`?
Or should I just hardcode the path inside the test assets artifact that contains the folders?

The trailing `/` will be fixed in Helix SDK later

directhex replied:
Suggestions from @akoeplinger on internal chat:

should be clearer with `<XHarnessAppFolderToTest Include="@(_appBundleDirs)" Condition="%(Identity.EndsWith('.app')) or %(Identity.EndsWith('.app/'))` I think
assuming these directories are at the edge of the tree, rather than somewhere in the middle
or even easier: `<XHarnessAppFolderToTest Include="$([System.IO.Directory]::GetDirectories('$(TestArchiveTestsRoot)', '*.app', System.IO.SearchOption.AllDirectories))" />`

premun committed:
Address some of Santiago's feedback from Jo's PR

premun committed:
Merge runtime.yml from master

premun committed:
Build iOSx64 libraries

premun committed:
Try to build mono product

premun committed:
Try to build mono product

premun committed:
Merge master

premun committed:
Do not build AllSubsets

premun committed:
Remove dead code

premun committed:
Revert: Do not built AllSubsets

premun committed:
Disable test archivation

directhex committed:
Improvements for local test running

directhex committed:
Merge branch 'prvysoky/use-helix-sdk' of github.com:premun/runtime into prvysoky/use-helix-sdk

premun committed:
Return dead code

directhex committed:
Reintroduce whitespace errors

None of these fixes are wrong per se (maybe the ApkBuilder changes?) but they complicate the PR by mixing unrelated changes

premun committed:
Bump Helix SDK and keep trailing slashes

premun committed:
Revert "Bump Helix SDK and keep trailing slashes"

This reverts commit d1a42261c34b5413f306ccaca37fed24294b8ac6.

premun committed:
Try building 2 app bundles

directhex committed:
Try a cleaner `.app` find in msbuild, from @akoeplinger

directhex committed:
Revert "Try a cleaner `.app` find in msbuild, from @akoeplinger"

This reverts commit 64da851be89f6f4dd00be7f804534aa3af15259d.

directhex committed:
Do not archiveTests on iOS, it causes tests to be included twice

directhex committed:
Revert "Revert "Try a cleaner `.app` find in msbuild, from @akoeplinger""

This reverts commit 506afd2c2b00800daf75b86f8ce48acb1c83ad84.

directhex committed:
Turn on all the tests on iOS

directhex committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

directhex committed:
Remove Android changes, to keep scope of PR smaller

premun commented on global.json:
```diff
@@ -15,7 +15,7 @@
     "Microsoft.DotNet.Arcade.Sdk": "5.0.0-beta.20280.1",
     "Microsoft.DotNet.Build.Tasks.SharedFramework.Sdk": "5.0.0-beta.20280.1",
-    "Microsoft.DotNet.Helix.Sdk": "5.0.0-beta.20280.1",
+    "Microsoft.DotNet.Helix.Sdk": "5.0.0-beta.20304.1",
```
Can we try this version?

```suggestion
    "Microsoft.DotNet.Helix.Sdk": "5.0.0-beta.20308.1",
```

premun commented on src/libraries/sendtohelix.proj:
```diff
@@ -40,6 +40,12 @@
+  <PropertyGroup Condition="'$(TargetsMobile)' == 'true'">
+    <ArchiveTests>false</ArchiveTests>
+    <IncludeDotNetCli>true</IncludeDotNetCli>
+    <MicrosoftDotNetXHarnessCLIVersion>1.0.0-prerelease.20304.1</MicrosoftDotNetXHarnessCLIVersion>
```
```suggestion
    <MicrosoftDotNetXHarnessCLIVersion>1.0.0-prerelease.20308.1</MicrosoftDotNetXHarnessCLIVersion>
```

directhex committed:
Update global.json

Co-authored-by: Přemek Vysoký <premek.vysoky@microsoft.com>

premun commented on src/libraries/sendtohelix.proj:
```diff
@@ -40,6 +40,12 @@
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(TargetsMobile)' == 'true'">
+    <ArchiveTests>false</ArchiveTests>
```
This ArchiveTests doesn't do anything I think. It was my attempt that didn't work and then you resolved it in yaml

directhex committed:
Update src/libraries/sendtohelix.proj

Co-authored-by: Přemek Vysoký <premek.vysoky@microsoft.com>

directhex committed:
Remove useless archiveTests=false

premun commented on src/libraries/sendtohelix.proj:
```diff
@@ -97,7 +103,17 @@
 
-    <ItemGroup>
+    <ItemGroup Condition="'$(TargetOS)' == 'iOS' or '$(TargetOS)' == 'tvOS'">
+      <!-- Find all directories named *.app and strip the trailing slash -->
```
```suggestion
      <!-- Find all directories named *.app -->
```

directhex committed:
Update src/libraries/sendtohelix.proj

Co-authored-by: Přemek Vysoký <premek.vysoky@microsoft.com>

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

premun committed:
Fix runtime-packs dir

premun commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -104,6 +104,13 @@ jobs:
+              - task: CopyFiles@2
+                displayName: Prepare lib-runtime-packs folder to publish
+                inputs:
+                  sourceFolder: $(Build.SourcesDirectory)/artifacts/bin/microsoft.netcore.app.runtime.ios-x64
```
This will need to be made dynamic so that it works for other platforms as well but I am not sure how it works and how it gets consumed

More information here: https://github.com/dotnet/runtime/commit/2b7852b3d039b7d40f2623a3ec54e7d23eb93790

ViktorHofer replied:
Opened https://github.com/dotnet/runtime/pull/37652 to unblock you.

safern commented on src/libraries/sendtohelix.proj:
```diff
@@ -48,7 +53,7 @@
 
   <PropertyGroup Condition="'$(HelixCommand)' == '' and '$(BuildAllConfigurations)' == 'true'">
-    <HelixPreCommands>set DOTNET_CLI_TELEMETRY_OPTOUT=1;set DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1;set DOTNET_MULTILEVEL_LOOKUP=0</HelixPreCommands>
+    <HelixPreCommands>$(HelixPreCommands); set DOTNET_CLI_TELEMETRY_OPTOUT=1;set DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1;set DOTNET_MULTILEVEL_LOOKUP=0</HelixPreCommands>
```
Why is this change needed? `BuildAllConfigurations` has nothing to do with iOS.

premun replied:
You're right, it's not

safern commented on src/libraries/sendtohelix.proj:
```diff
@@ -40,6 +40,11 @@
 
+  <PropertyGroup Condition="'$(TargetsMobile)' == 'true'">
+    <IncludeDotNetCli>true</IncludeDotNetCli>
+    <MicrosoftDotNetXHarnessCLIVersion>1.0.0-prerelease.20308.1</MicrosoftDotNetXHarnessCLIVersion>
```
Use the property in Versions.props instead?

premun replied:
There was none, but I added it.

safern committed:
Revert changes in yml to use new mono build workflow

safern committed:
Merge remote-tracking branch 'dotnet/master' into prvysoky/use-helix-sdk

safern committed:
Add iOS x64 testing to new collapsed build model

safern committed:
Just run System.Buffers.Tests to make testing faster

safern committed:
Set helixQueuesTemplate in the right place

premun committed:
Set XHarness version in Version.props

# Conflicts:
#	src/libraries/sendtohelix.proj

premun committed:
Roll back to XHarness that supports Xcode 11.4

premun commented on src/libraries/tests.proj:
```diff
@@ -19,7 +19,7 @@
 
   <ItemGroup>
     <!-- We currently only test with C# projects. -->
-    <ProjectReference Include="$(MSBuildThisFileDirectory)*\tests\**\*.Tests.csproj"
```
@safern won't this influence other platforms as well? Maybe we want to add a special clause for `TargetsMobile`?

safern committed:
Cleanup

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

# Conflicts:
#	eng/Versions.props

premun committed:
Add XHarness CLI to Version.Details.xml

premun committed:
Bump XHarness and Helix SDK versions

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

premun committed:
Target iOS 13.5 specifically

premun committed:
Bump XHarness CLI version

directhex committed:
Merge remote-tracking branch 'origin/master' into enable-ios-helix

directhex committed:
Minor helix tweak maybe

directhex committed:
Ensure xharness is included for iOS

directhex committed:
Enable all tests

directhex committed:
Remove duplicate

directhex committed:
Add exclude list of test suites without 100% pass on iOS

directhex committed:
Exclude more tests

directhex committed:
Skip more tests

directhex committed:
Merge remote-tracking branch 'origin/master' into enable-ios-helix

directhex committed:
Exclude more tests

directhex commented:
Should have been reading the errors in more detail

Looks like most of our errors right now are from Helix:

```
[07:22:51] dbug: test[0]
      07:22:51.3221030 4382 Execution timed out after 30 seconds and the process was killed.
[07:22:51] dbug: test[0]
      07:22:51.3269460 Process exited with 137
[07:22:51] dbug: test[0]
      07:22:51.3281860 mlaunch failed when listing simulators but trying to parse the results anyway
[07:22:51] dbug: test[0]
      07:22:51.3352210 Failed to find/create simulator (attempt 1/3):
      System.Exception: Failed to list simulators - no XML with devices found. mlaunch timed out with 137)
         at Microsoft.DotNet.XHarness.iOS.Shared.Hardware.SimulatorLoader.LoadDevices(ILog log, Boolean includeLocked, Boolean forceRefresh, Boolean listExtraData) in /_/src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:line 80
         at Microsoft.DotNet.XHarness.iOS.Shared.Hardware.SimulatorLoader.FindOrCreateDevicesAsync(ILog log, String runtime, String devicetype, Boolean force) in /_/src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:line 192
         at Microsoft.DotNet.XHarness.iOS.Shared.Hardware.SimulatorLoader.FindSimulators(TestTargetOs target, ILog log, Boolean createIfNeeded, Boolean minVersion) in /_/src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:line 385
         at Microsoft.DotNet.XHarness.iOS.AppRunner.RunApp(AppBundleInformation appInformation, TestTargetOs target, TimeSpan timeout, TimeSpan testLaunchTimeout, String deviceName, String companionDeviceName, Boolean ensureCleanSimulatorState, Int32 verbosity, XmlResultJargon xmlResultJargon, String[] skippedMethods, String[] skippedTestClasses, CancellationToken cancellationToken) in /_/src/Microsoft.DotNet.XHarness.iOS/AppRunner.cs:line 115
[07:22:51] dbug: test[0]
      07:22:51.3401080 /Applications/Xcode115.app/Contents/Developer/usr/bin/simctl create "iPhone-X (iOS-13-5) - created by XHarness" com.apple.CoreSimulator.SimDeviceType.iPhone-X com.apple.CoreSimulator.SimRuntime.iOS-13-5
```

premun commented:
@directhex I noticed this issue a while ago and logged it here: https://github.com/dotnet/xharness/issues/283

mlaunch fails to get a simulator so I added a retry (notice the `attempt 1/3` which mostly solved the symptom but as I currently don't have any time dedicated to XHarness I didn't solve the underlying issue.

I will try to get some time as this will be a blocker soon and I need to investigate if it's machine-specific issue

premun commented on eng/testing/AppleRunnerTemplate.sh:
```diff
@@ -38,15 +35,25 @@ while true; do
         sleep 5
     fi
 done
+# Restart the simulator to make sure it is tied to the right user session
+xcode_version=11.4
+xcode_path="/Applications/Xcode${xcode_version/./}.app"
+simulator_app="$xcode_path/Contents/Developer/Applications/Simulator.app"
+pid=`ps aux | grep "$simulator_app" | grep -v grep | tr -s ' ' | cut -d ' ' -f 2`
+if [ ! -z "$pid" ]; then
+    sudo kill "$pid"
+fi
+open -a "$simulator_app"
```
```suggestion

# Restart the simulator to make sure it is tied to the right user session
xcode_version=11.4
xcode_path="/Applications/Xcode${xcode_version/./}.app"
simulator_app="$xcode_path/Contents/Developer/Applications/Simulator.app"
sudo pkill -9 -f "$simulator_app"
open -a "$simulator_app"
```

premun commented on eng/testing/AppleRunnerTemplate.sh:
```diff
@@ -38,15 +35,25 @@ while true; do
 _exitCode=$?
 
-echo "Xharness artifacts: $XHARNESS_OUT"
+echo "Xharness artifacts: `ls -lh $XHARNESS_OUT`"
```
```suggestion
echo "XHarness artifacts: `ls -lh $XHARNESS_OUT`"
```

premun commented on eng/testing/AppleRunnerTemplate.sh:
```diff
@@ -38,15 +35,25 @@ while true; do
-
-dotnet xharness ios test --app="$APP_BUNDLE" \
-    --targets=$TARGET \
+dotnet xharness ios test \
```
Here I would also recommend specifying `--timeout` and `--launch-timeout`.
The former is for the overall run, the latter for how long it takes to boot the simulator, install the app and start it.

The failures you are seeing (mlaunch fails to find a simulator) have a retry logic and can take time from what I've seen so before we fix it, it could help increasing the timeouts (say 15 minutes for the launch timeout and 30-40 for the overall timeout)

premun reviewed (commented):
@directhex I noticed this issue a while ago and logged it here: https://github.com/dotnet/xharness/issues/283

mlaunch fails to get a simulator so I added a retry (notice the `attempt 1/3` which mostly solved the symptom but as I currently don't have any time dedicated to XHarness I didn't solve the underlying issue.

I will try to get some time as this will be a blocker soon and I need to investigate if it's machine-specific issue

premun committed:
Kill the simulators better

premun committed:
Merge branch 'master' into prvysoky/use-helix-sdk

Conflicts:
	eng/pipelines/common/platform-matrix.yml
	src/libraries/sendtohelixhelp.proj

premun committed:
Re-use the XCODE_PATH variable, unify variable naming

premun committed:
Rename XHarnessAppFolderToTest to match new API

premun committed:
Do not kill simulator after the run

directhex commented:
@premun could this be something as stupid as not giving mlaunch enough time to finish?

Here's a good run:

```
[05:25:47] dbug: test[0]
      05:25:47.1394590 /private/tmp/helix/working/A9DE099B/p/microsoft.dotnet.xharness.cli/1.0.0-prerelease.20352.3/tools/netcoreapp3.1/any/../../../runtimes/any/native/mlaunch/bin/mlaunch --sdkroot /Applications/Xcode115.app --listsim=/tmp/tmpmlrVXC.tmp --output-format=XML
[05:26:08] dbug: test[0]
      05:26:08.8570810 Simulator configuration written to /tmp/tmpmlrVXC.tmp
[05:26:08] dbug: test[0]
      05:26:08.9515020 Process exited with 0
[05:26:09] dbug: test[0]
      05:26:08.9536870 Simulator listing returned:
```

That took 22 seconds to return any results

Here's a bad one:

```
[02:06:49] dbug: test[0]
      02:06:49.7723290 /private/tmp/helix/working/A9DE099B/p/microsoft.dotnet.xharness.cli/1.0.0-prerelease.20352.3/tools/netcoreapp3.1/any/../../../runtimes/any/native/mlaunch/bin/mlaunch --sdkroot /Applications/Xcode115.app --listsim=/tmp/tmp3dqcCa.tmp --output-format=XML
[02:07:19] dbug: test[0]
      02:07:19.8645850 Process 4094 didn't exit within 00:00:30 and will be killed
```

Timeout is 30 seconds

How close to the wire is it getting?

premun commented:
@directhex I have observed these behaviours so far:
- mlaunch takes about 1 second to list simulators successfully
- mlaunch takes really long and times out
- mlaunch takes really long and times out **but** lists simulators successfully
- mlaunch returns a non-0 exit code
- mlaunch returns a non-0 exit code **but** lists simulators successfully

What I did is implement a retry logic to give 3 attempts to this and also even when mlaunch fails, I try to read the XML files with the results so sometimes mlaunch fails and I still get the list of simulators.

I haven't had much time investigating this though so in order to unblock us, I will need to talk to my manager first.

In the meantime I did some experimentation on one of the machines where I see a lot of failures but it is quite hard to reproduce sometimes...

premun commented:
I have an attempt to fix the issue: https://github.com/dotnet/xharness/pull/288
So we will see..

premun commented:
/azprun

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

premun committed:
Fix counting of Helix work items

premun committed:
Increase timeouts a bit and remove the timeout in seconds

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

premun committed:
Do not set Helix timeout, XHarness SDK will do it for us

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

premun commented:
@directhex I have fixed the mlaunch issues and we can move forward with this if you wish.

I rebased and we can see how the tests go.

After we merge the newest Helix SDK (after an Arcade rollout), the `<LaunchTimeout>` value will also come into effect and improve edge cases where simulators would take long to boot. However, I don't expect this to happen after the fix I made to the mlaunch problem.

premun commented:
So it seems the issue is still on-going, will revisit tomorrow.

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

premun committed:
Bump the Helix SDK

premun committed:
Bump the Helix SDK

premun committed:
Exclude Microsoft.Extensions.Logging.Console.Tests because of failing Log_WritingScopes_LogsWithCorrectColorsWhenColorEnabled

premun committed:
Merge remote-tracking branch 'origin/master' into prvysoky/use-helix-sdk

premun committed:
Merge remote-tracking branch 'dotnet/master' into prvysoky/use-helix-sdk

premun committed:
Bump XHarness CLI and Helix SDK

premun committed:
Merge remote-tracking branch 'dotnet/master' into prvysoky/use-helix-sdk

Conflicts:
	eng/Version.Details.xml
	eng/Versions.props
	eng/pipelines/common/platform-matrix.yml
	global.json

directhex committed:
Merge remote-tracking branch 'origin/master' into enable-ios-helix

directhex commented:
Just doing one sanity build, to make sure it wasn't a fluke that it passed last time, but given last time was a glorious pass, I am feeling optimistic

premun commented:
@directhex I have run about 20 builds and the **Send to Helix** part was 100% fine. The Build step however failed occasionally. It would take 2 hours and then timeout. The build seemed to keep on building, just slowly.

I have also observed this on my local Mac Mini - building the iOS apps took waaay longer than I have seen in AzDO.

directhex commented:
@premun I've actually just been handed a bug about build timeouts on a completely unrelated job - https://github.com/dotnet/runtime/issues/40933

Infra issue?

directhex committed:
Re-enable all lanes, so merging is viable

directhex committed:
Merge remote-tracking branch 'origin/master' into enable-ios-helix

directhex commented:
Wait, can it be infra if it's happening on both Mac and Linux? Like, Mac is hosted hardware, Linux is Azure VMs, they couldn't be more different yet both are showing this strange slowdown behaviour. New CoreCLR bug?

premun commented:
@directhex I was observing this already few months ago. No idea whether it's tied to a specific machine too since re-runs usually help.

directhex committed:
Merge remote-tracking branch 'origin/master' into enable-ios-helix

directhex commented:
Hm, seems to legit break the Wasm lane

premun commented:
> Hm, seems to legit break the Wasm lane

@directhex I will have a look

directhex commented:
@premun I can repro the issue locally.

directhex committed:
Fix Wasm breakage

directhex commented:
Resolved.

directhex committed:
Merge remote-tracking branch 'origin/master' into enable-ios-helix

akoeplinger commented on eng/testing/AppleRunnerTemplate.sh:
```diff
@@ -39,14 +36,22 @@ while true; do
-    --targets=$TARGET \
+# Restart the simulator to make sure it is tied to the right user session
+XCODE_VERSION=11.4
+XCODE_PATH="/Applications/Xcode${XCODE_VERSION/./}.app"
```
hardcoding the Xcode version will make it difficult to keep this updated when new Xcode versions are rolled out, or locally where I just have Xcode.app.

why do we not use `/Applications/Xcode.app`?

premun replied:
We have several versions installed side-by-side on our on-prem machines. It is a practice adopted from Xamarin as it is then super clear on which version is used.

You could change this script to get the version from `xcode-select -p`

akoeplinger replied:
Yes using `xcode-select` would be much better.

premun replied:
I will give it a go, and also will stop specifying the iOS version since if we install a new Xcode, it might not have 13.5 available

premun replied:
We basically won't be able to target one version of iOS with confidence.. We can also not specify the version and XHarness will grab the latest

premun committed:
Use xcode-select for local runs

premun committed:
Target the most recent iOS version

premun committed:
Do not kill simulator in local dev env

premun committed:
Fix comments

premun committed:
Merge remote-tracking branch 'dotnet/master' into prvysoky/use-helix-sdk

akoeplinger commented on eng/testing/AppleRunnerTemplate.sh:
```diff
@@ -39,14 +35,17 @@ while true; do
 _exitCode=$?
 
-echo "Xharness artifacts: $XHARNESS_OUT"
+echo "XHarness artifacts: `ls -lh $XHARNESS_OUT`"
```
I feel like I commented about this in some PR already but I don't like the `ls` here. It means I can't cmd+click the path in my terminal to open the folder anymore since the path is no longer printed.

premun replied:
Changed

premun committed:
Print the output path instead of listing the files

akoeplinger reviewed (commented):
LGTM!

directhex commented:
https://helix.dot.net/api/2019-06-17/jobs/ba701e37-fa40-4a78-92d9-729043b36d9c/workitems/Microsoft.Extensions.Configuration.Json.Tests/console

Timed out installing the test app?

directhex committed:
Merge branch 'master' into prvysoky/use-helix-sdk

safern commented on eng/testing/AppleRunnerTemplate.sh:
```diff
@@ -39,14 +35,17 @@ while true; do
 
-dotnet xharness ios test --app="$APP_BUNDLE" \
-    --targets=$TARGET \
+dotnet xharness ios test  \
```
Will we support filtering tests or any xunit parameters? Like `-notrait Category=Outerloop`? If so, this is wrong and should use `RunScriptCommand` in the MSBuild side to set this command.

akoeplinger replied:
Ideally yes, but we first need a way to flow arbitrary command line arguments to the iOS/Android test apps and that will need some changes in xharness and the iOS/Android test runners, I don't think we need to block the PR on that.

safern replied:
Sounds good. We just need to be conscious that PR validation is running outerloop tests for iOS as well, which might be causing the slow runs.

akoeplinger replied:
ah no, we added handling for that already with https://github.com/dotnet/runtime/blob/3c1b48ee29f13cac93a86829374b917375a49a2d/eng/testing/tests.mobile.targets#L76 which is read by the testrunner so Outerloop and failing tests should already be excluded.

safern replied:
Ok, sounds good. Maybe we should do it how we do it in other places passing them as an argument but that is in the future, no need to block this PR.

safern commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -24,13 +24,11 @@
+        '$(Scenario)' == 'gcstress0xc_jitminopts_heapverify1'">01:30:00</_workItemTimeout>
+    <_workItemTimeout Condition="'$(Scenario)' == '' and '$(_workItemTimeout)' == '' and ('$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm')">00:45:00</_workItemTimeout>
+    <_workItemTimeout Condition="'$(Scenario)' != '' and '$(_workItemTimeout)' == '' and ('$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm')">01:00:00</_workItemTimeout>
+    <_workItemTimeout Condition="'$(Scenario)' == '' and '$(_workItemTimeout)' == ''">00:30:00</_workItemTimeout>
```
30 minutes? That is too much, not? Should we set this special timeout for iOS only? We used to have 15 mins timeout.

premun replied:
I reset the previous ones and set one specifically for iOS/tvOS. I, however, don't understand what the `Scenario` is for and whether I should also have one for non-empty `Scenario` and iOS/tvOS?

safern replied:
Nope. Scenarios are for special test modes that are translated into a job that sets an environment variable before running the tests.

At the moment there is no such thing for iOS/tvOS, so let's just set it for empty scenario and iOS/tvOS. If the need comes later on we can adjust.

safern commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -24,13 +24,11 @@
+    <_workItemTimeout Condition="'$(Scenario)' == '' and '$(_workItemTimeout)' == '' and ('$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm')">00:45:00</_workItemTimeout>
+    <_workItemTimeout Condition="'$(Scenario)' != '' and '$(_workItemTimeout)' == '' and ('$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm')">01:00:00</_workItemTimeout>
+    <_workItemTimeout Condition="'$(Scenario)' == '' and '$(_workItemTimeout)' == ''">00:30:00</_workItemTimeout>
+    <_workItemTimeout Condition="'$(Scenario)' != '' and '$(_workItemTimeout)' == ''">00:45:00</_workItemTimeout>
```
Ditto.

premun commented:
> https://helix.dot.net/api/2019-06-17/jobs/ba701e37-fa40-4a78-92d9-729043b36d9c/workitems/Microsoft.Extensions.Configuration.Json.Tests/console
> 
> Timed out installing the test app?

Looks like it happened again but on the very same machine (**dci-macpro-06**). I will investigate what's up

premun committed:
Merge remote-tracking branch 'dotnet/master' into prvysoky/use-helix-sdk

premun committed:
Reset timeouts and only change it for iOS/tvOS

premun committed:
Merge remote-tracking branch 'dotnet/master' into prvysoky/use-helix-sdk

premun commented:
I took the problematic machine offline for investigations so let's see whether the rest goes green now.
I will have time to investigate next week.

premun commented:
Alright, looks everything is green. So I wonder if we want to check this in or not. The reason is that sometimes the iOS build times out when it is on one of the slow agents (or whatever the reason is), so this will bring flakey-ness. On the other hand, I am not sure whether this is not happening already then?

safern commented:
I think if we don't check this in we won't find those flaky issues and fix them. However, the question I have is, do we have the capacity yet to enable this on all PRs and CI?

premun commented:
The capacity for the iOS tests itself is OK. The tests only take 6 minutes right now. We have 59 machines for that and there's 155 of work items so the test itself is usually super short. The capacity for building the iOS tests is coming from AzDO and I am not sure how are we doing there. I will try to find out.

premun committed:
Merge remote-tracking branch 'dotnet/master' into prvysoky/use-helix-sdk

safern commented:
Building the product and tests took 36 mins in last build which is not bad.

directhex commented:
Practically speaking, what blockers remain on this PR?

premun commented:
> @directhex Practically speaking, what blockers remain on this PR?

So, from my point of view (and please consider that this is my first and only interaction with `dotnet/runtime` so far and I don't have experience in how the overall PR builds behave):
- Building the iOS apps times out every now and then when a "slow agent" is processing the build. A re-run is needed. I am not sure whether this causes problems with other legs already since it doesn't seem to be tied with this particular leg but..
- I don't know how often it happens on regular basis that you need to re-run a leg in a PR build so I don't know what is the "regular flakey-ness".
- This week one machine somehow went bad and installation of apps started to hang. I took it offline for investigations but I am not sure about the root cause still and I am not sure this won't happen again to bring more flakey-ness. (but as @safern noted, maybe the only way to find out is to check it in and observe).
- I am not sure how to approach the question of capacity. I personally think we are fine for iOS but again - maybe the only way to see is to check it in.

So.. I don't know how flakey the PR builds in `dotnet/runtime` are already and whether we won't bring a lot of pain by checking this in, but I guess this is up to you guys. I don't know whether we shouldn't just flag the leg as not required and observe first or what is the regular procedure when adding new platforms, but I just don't have experience with this repository to decide this.

Other than that, the iOS tests part seems quite stable and fast, so..

safern commented:
> Building the iOS apps times out every now and then when a "slow agent" is processing the build. A re-run is needed. I am not sure whether this causes problems with other legs already since it doesn't seem to be tied with this particular leg but..

We're already seeing this in other OSX, iOS legs and it seems to be an AzDo bug which seems to be under investigation so I wouldn't be concerned by this and also seems like it has somewhat being mitigated.

> This week one machine somehow went bad and installation of apps started to hang. I took it offline for investigations but I am not sure about the root cause still and I am not sure this won't happen again to bring more flakey-ness. (but as @safern noted, maybe the only way to find out is to check it in and observe).

I still stand my point the best way to mitigate issues is to check-in and investigate issues that come show up.

> I am not sure how to approach the question of capacity. I personally think we are fine for iOS but again - maybe the only way to see is to check it in.

Given the number of machines you shared above and the time it is taking to process work items it seems fine. We won't know this with certainty until we check it in, and as we enable more test projects and run on more PRs if we hit an issue we can always disable the leg as we did for a couple of days for WASM until we got more capacity. 

Overall, this LGTM and we should try checking it in, specially since master is already 6.0.

steveisok commented:
@premun Please give your approval on the PR and we can merge.

ViktorHofer commented on eng/testing/AppleRunnerTemplate.sh:
```diff
@@ -1,11 +1,9 @@
-TARGET_ARCH=$2
-TARGET=
-SCHEME_SDK=
+[[RunCommands]]
```
I'm trying to get rid of the runcommands templating mechanism as having individual runner scripts for different targets (ios/android/wasm) seems to be sufficient. Doing that in https://github.com/dotnet/runtime/pull/39923. That will basically undo this change here. Any objections?

safern replied:
No, I support that change as we should be able to define it via MSBuild.

steveisok replied:
No objections from me.
